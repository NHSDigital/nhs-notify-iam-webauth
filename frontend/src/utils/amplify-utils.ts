/* eslint-disable import/no-unresolved,@typescript-eslint/no-require-imports */
/* eslint-disable unicorn/prefer-module */
/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable @typescript-eslint/no-require-imports */
import { cookies } from 'next/headers';
import { createServerRunner } from '@aws-amplify/adapter-nextjs';
import { fetchAuthSession } from 'aws-amplify/auth/server';

const config = require('@/amplify_outputs.json');

export const { runWithAmplifyServerContext } = createServerRunner({
  config,
});

export async function getAccessTokenServer(): Promise<string | undefined> {
  try {
    const { tokens } = await runWithAmplifyServerContext({
      nextServerContext: { cookies },
      operation: fetchAuthSession,
    });

    return tokens?.accessToken?.toString();
  } catch {
    // no-op
  }
}
