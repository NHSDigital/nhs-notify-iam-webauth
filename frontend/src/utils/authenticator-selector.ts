import { useAuthenticator } from '@aws-amplify/ui-react';

type UseAuthenticatorSelector = Exclude<
  Parameters<typeof useAuthenticator>[0],
  undefined
>;

export const authenticatorSelector: UseAuthenticatorSelector = (ctx) => [
  ctx.authStatus,
];
