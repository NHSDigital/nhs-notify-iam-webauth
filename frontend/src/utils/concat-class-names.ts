type ClassName = string | undefined | false;

export default function concatClassNames(...classNames: ClassName[]): string {
  return classNames
    .filter(Boolean)
    .join(' ')
    .replaceAll(/\s{2,}/g, ' ')
    .trim();
}
