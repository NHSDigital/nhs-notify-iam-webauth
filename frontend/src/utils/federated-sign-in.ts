import { signInWithRedirect } from '@aws-amplify/auth';
// eslint-disable-next-line import-x/no-unresolved
import amplifyOutputs from '@/amplify_outputs.json';

export interface State {
  redirectPath: string;
}

export function federatedSignIn(redirectPath: string) {
  const providers = amplifyOutputs.auth?.oauth?.identity_providers || [];
  if (providers.length !== 1) {
    throw new Error('Invalid OAUTH provider configuration');
  }
  const provider = providers[0];
  if (!provider) {
    throw new Error('Missing OAUTH custom provider configuration');
  }
  return signInWithRedirect({
    provider: {
      custom: provider,
    },
    customState: JSON.stringify({ redirectPath }),
  });
}
