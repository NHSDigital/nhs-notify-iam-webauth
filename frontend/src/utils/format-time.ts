export const formatTime = (remainingSeconds: number) => {
  const remainingMinutes = Math.round(remainingSeconds / 60);

  if (remainingMinutes <= 0) {
    return `${remainingSeconds} seconds`;
  }

  if (remainingMinutes <= 1) {
    return '1 minute';
  }

  return `${remainingMinutes} minutes`;
};
