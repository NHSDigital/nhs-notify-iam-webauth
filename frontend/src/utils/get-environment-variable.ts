export const getEnvironmentVariable = (name: string) => {
  const secret = process.env[name];

  if (!secret) {
    throw new Error('Missing environment variable');
  }

  return secret;
};
