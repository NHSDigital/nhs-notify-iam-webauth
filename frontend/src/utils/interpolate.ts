/**
 * Replaces variables within a template string using a variables object, and supports pluralisation.
 *
 * Supports:
 * - simple variable interpolation: `{{name}}`
 * - pluralisation: `{{variable|singular|plural}}`
 *
 * Example:
 * interpolate('Hello {{name}}, you have {{count}} {{count|message|messages}}.', {
 *   name: 'John',
 *   count: 2,
 * }); // => 'Hello John, you have 2 messages.'
 */

// eslint-disable-next-line security/detect-unsafe-regex, sonarjs/slow-regex
const interpolationPattern = /{{([^|}]+)(?:\|([^|}]+)\|([^}]+))?}}/g;

export function interpolate(
  template: string,
  variables: Record<string, string | number> = {}
): string {
  // eslint-disable-next-line unicorn/prefer-string-replace-all
  return template.replace(interpolationPattern, (_, key, singular, plural) => {
    const value = variables[key];

    if (singular !== undefined && plural !== undefined) {
      const numeric = Number(value);

      if (Number.isNaN(numeric)) return plural;
      return numeric === 1 ? singular : plural;
    }

    return value == null ? '' : String(value);
  });
}
