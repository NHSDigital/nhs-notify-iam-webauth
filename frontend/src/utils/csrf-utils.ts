'use server';

import { createHmac, randomBytes } from 'node:crypto';
import { cookies } from 'next/headers';
import { logger } from '@nhs-notify-iam-webauth/utils-logger';
import { getEnvironmentVariable } from './get-environment-variable';
import { getSessionId } from './amplify-utils';

export const generateSessionCsrfToken = async (sessionId: string) => {
  const secret = getEnvironmentVariable('CSRF_SECRET');

  const salt = randomBytes(8).toString('hex');

  const hash = createHmac('sha256', secret)
    .update(sessionId)
    .update(salt)
    .digest('hex');

  const csrfToken = `${hash}.${salt}`;

  return csrfToken;
};

export const verifyCsrfToken = async (
  csrfToken: string,
  secret: string,
  sessionId: string
) => {
  const [hmac, salt] = csrfToken.split('.');

  const expectedHmac = createHmac('sha256', secret)
    .update(sessionId)
    .update(salt)
    .digest('hex');

  return expectedHmac === hmac;
};

export const verifyFormCsrfToken = async (formData: FormData) => {
  const secret = getEnvironmentVariable('CSRF_SECRET');
  const sessionId = await getSessionId();

  if (!sessionId) {
    logger.error('Unauthenticated');
    return false;
  }

  const cookieStore = await cookies();
  const csrfTokenCookie = cookieStore.get('csrf_token');

  if (!csrfTokenCookie) {
    logger.error('missing CSRF cookie');
    return false;
  }

  const csrfTokenCookieValue = csrfTokenCookie.value;

  const csrfTokenFormField = formData.get('csrf_token');

  if (!csrfTokenFormField) {
    logger.error('missing CSRF form field');
    return false;
  }

  if (csrfTokenFormField !== csrfTokenCookieValue) {
    logger.error('CSRF mismatch');
    return false;
  }

  const formVerification = await verifyCsrfToken(
    csrfTokenFormField,
    secret,
    sessionId
  );

  if (!formVerification) {
    logger.error('CSRF error');
    return false;
  }

  return true;
};
