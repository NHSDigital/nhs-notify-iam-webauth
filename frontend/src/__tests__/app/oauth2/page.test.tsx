import { act, render, waitFor } from '@testing-library/react';
import { redirect } from 'next/navigation';
import { Hub } from 'aws-amplify/utils';
import OAuth2CallbackPage from '@/src/app/oauth2/page';

jest.mock('aws-amplify/utils', () => ({
  Hub: {
    listen: jest.fn(),
  },
}));

jest.mock('next/navigation', () => ({
  redirect: jest.fn(),
  RedirectType: { replace: 'replace' },
}));

const mockedHub = jest.mocked(Hub);
const mockedRedirect = jest.mocked(redirect);

function getEventListener() {
  const lastHubListenCall = mockedHub.listen.mock.lastCall;
  const eventType = lastHubListenCall?.[0];
  const eventListener = lastHubListenCall?.[1];

  expect(eventType).toBe('auth');

  expect(eventListener).toBeTruthy();
  return eventListener!;
}

it('listens for oauth state events from amplify hub', async () => {
  const page = <OAuth2CallbackPage />;
  const container = render(page);

  const eventListener = getEventListener();

  act(() => {
    eventListener({
      payload: {
        event: 'customOAuthState',
        data: '{"redirectPath":"/testing"}',
      },
      channel: '',
    });
    container.rerender(page);
  });

  await waitFor(() =>
    expect(mockedRedirect).toHaveBeenCalledWith(
      '/signin?redirect=%2Ftesting',
      'replace'
    )
  );
});

it('ignores other hub events', async () => {
  const page = <OAuth2CallbackPage />;
  const container = render(page);

  const eventListener = getEventListener();

  act(() => {
    eventListener({
      payload: {
        event: 'ignoreThis',
        data: '{"redirectPath":"/testing"}',
      },
      channel: '',
    });
    container.rerender(page);
  });

  await waitFor(() => expect(mockedRedirect).not.toHaveBeenCalled());
});

it('defaults redirect path', async () => {
  const page = <OAuth2CallbackPage />;
  const container = render(page);

  const eventListener = getEventListener();

  act(() => {
    eventListener({
      payload: {
        event: 'customOAuthState',
        data: '{"redirectPath":""}',
      },
      channel: '',
    });
    container.rerender(page);
  });

  await waitFor(() =>
    expect(mockedRedirect).toHaveBeenCalledWith(
      '/signin?redirect=%2Ftemplates%2Fmanage-templates',
      'replace'
    )
  );
});
