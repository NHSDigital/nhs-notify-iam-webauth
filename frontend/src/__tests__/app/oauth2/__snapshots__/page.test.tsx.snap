// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CIS2CallbackPage should redirect to signin page with custom redirect when user is authenticated, renders loading spinner 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-loader"
  >
    <span
      class="nhsuk-loader__spinner"
    />
    <h1
      class="nhsuk-loader__text"
    >
      Redirecting, please wait
    </h1>
  </div>
</DocumentFragment>
`;
